### PR TITLE
Tetrahedral orientations [ho-tet-ori-dev]

### PR DESCRIPTION
**Pull request by mlstowell**
Our current algorithm for adaptive mesh refinement on tetrahedral meshes involves permuting the vertices of each tetrahedron so that the first edge is the longest.  If the mesh coordinates are defined by a `GridFunction`, the DoFs of this `GridFunction` must also be permuted.  This permutation was not supported when the element contained more than one internal degree of freedom i.e. for H1 elements with order greater than four or for L2 elements of any order.

This PR defines the 24 possible permutations of a tetrahedron as well as the algorithms to permute interior DoFs for H1 and L2 tetrahedra of arbitrary order.  It also adds a simple mesh generation miniapp called `twist` which generates meshes suitable for testing this new capability.  The `twist` miniapp may also provide the first example or miniapp with a means of demonstrating MFEM's strategy for handling periodic domains.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1306](https://github.com/mfem/mfem/pull/1306) | @mlstowell | @tzanio | @YohannDudouit + @aschaf | 02/22/20 | 03/12/20 | 03/22/20 | |

### CODE REVIEW

**Comment by tzanio** [comment]
Can you address the Travis and Appveyor failures please?

**Comment by mlstowell** [comment]
@tzanio , I added another unit test as well.  It's now passing the tests so I think it's ready to go.

**Comment by tzanio** [comment]
@YohannDudouit, do you need more time to review this?

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: miniapps/meshing/twist.cpp [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+//
+//             ----------------------------------------------
+//             Twist Miniapp:  Generate simple twisted meshes
+//             ----------------------------------------------
```
Comment (tzanio) [inline miniapps/meshing/twist.cpp]:
```suggestion
//          -------------------------------------------------------
//          Twist Miniapp:  Generate simple twisted periodic meshes
//          -------------------------------------------------------
```

Comment (tzanio) [reply miniapps/meshing/twist.cpp]: Mention this miniapp in `doc/CodeDocumentation.dox` and `CHANGELOG`

File: miniapps/meshing/twist.cpp [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+#include <fstream>
+#include <iostream>
+
+using namespace std;
+using namespace mfem;
+
+static Element::Type el_type_ = Element::WEDGE;
+static int    order_ = 3;
+static int    nz_    = 3;
+static int    ns_    = 0;
```
Comment (tzanio) [inline miniapps/meshing/twist.cpp]: Isn't it better if the default for this is not `0`? Also, should the name be "number of twists"?

File: miniapps/meshing/twist.cpp [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+static double c_     = 3.0;
+
+void pts(int iphi, int t, double x[]);
+void trans(const Vector &x, Vector &p);
+
+int main(int argc, char *argv[])
+{
+   int ser_ref_levels = 0;
+   int el_type = 8;
+   bool per_mesh = false;
```
Comment (tzanio) [inline miniapps/meshing/twist.cpp]: Probably better if the default is periodic=true?

**Comment by tzanio** [comment]
LGTM

File: miniapps/meshing/twist.cpp (hunk 2 above)
Comment (mlstowell) [reply miniapps/meshing/twist.cpp]: Sure I can change that.

**Pull request merged**